include/envoy/network/filter.h

```diff
@@ -227,6 +227,11 @@ class FilterManager {
    */
   virtual void addReadFilter(ReadFilterSharedPtr filter) PURE;
 
+  /**
+   * Remove a read filter from the connection.
+   */
+  virtual void removeReadFilter(ReadFilterSharedPtr filter) PURE;
```

Comment:
What is the use case for this one? Naively I would assume that we can just make a raw connection without any read filters, detect ALPN, and then attach an HTTP filter, etc. and continue?

Reply:
the tcp pool by default adds a read filter, and disconnects if data is read when no session is associated.
We could skip that and the add-and-remove but I'd want something to make sure early data didn't go into the void, so I'm mildly inclined to leave as-is.

Reply:
Ah OK. I see. I guess it would be nice to avoid this interface change but up to you.